Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -14,38 +15,179 @@ ComponentInfo::ComponentInfo()
 
 ComponentInfo::ComponentInfo(
     boost::shared_ptr<const std::vector<size_t>> assemblySortedDetectorIndices,
-    boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>> ranges)
+    boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+        detectorRanges,
+    boost::shared_ptr<const std::vector<size_t>> assemblySortedComponentIndices,
+    boost::shared_ptr<const std::vector<std::pair<size_t, size_t>>>
+        componentRanges,
+    boost::shared_ptr<std::vector<Eigen::Vector3d>> positions,
+    boost::shared_ptr<std::vector<Eigen::Quaterniond>> rotations,
+    boost::shared_ptr<DetectorInfo> detectorInfo)
     : m_assemblySortedDetectorIndices(std::move(assemblySortedDetectorIndices)),
-      m_ranges(std::move(ranges)),
-      m_size(m_assemblySortedDetectorIndices->size() + m_ranges->size()) {}
+      m_assemblySortedComponentIndices(
+          std::move(assemblySortedComponentIndices)),
+      m_detectorRanges(std::move(detectorRanges)),
+      m_componentRanges(std::move(componentRanges)),
+      m_positions(std::move(positions)), m_rotations(std::move(rotations)),
+      m_size(m_assemblySortedDetectorIndices->size() +
+             m_detectorRanges->size()),
+      m_detectorInfo(detectorInfo) {
+  if (m_rotations->size() != m_positions->size()) {
```

Comment:
If `ComponentInfo` is always created by the visitor, maybe we could consider removing all those checks?